packages/blocks/shuffle/src/components/add-entities-dialog.tsx

```diff
@@ -0,0 +1,205 @@
+  graphService?: GraphBlockHandler | null;
+}
+
+const _AddEntitiesDialog: ForwardRefRenderFunction<
+  AddEntitiesDialogRef,
+  AddEntitiesDialogProps
+> = ({ entityTypes, blockEntityId, onAddEntityItems, graphService }, ref) => {
+  const [open, setOpen] = useState(false);
+  const [selectedEntityType, setSelectedEntityType] = useState<EntityType>();
+  const [entityList, setEntityList] = useState<Entity[]>([]);
+  // `selections` is a object that stores checkbox `checked` state for each entity.
+  const [selections, setSelections] = useState<boolean[]>([]);
+
+  const handleClose = () => setOpen(false);
+
+  /**
+   * Defined a custom ref, so the parent can simply call `dialogRef.show()`,
+   * which also resets the state to it's initial value when modal is opened,
+   * instead of using an `useEffect` that listens modal's state
+   */
+  useImperativeHandle(
+    ref,
+    () => ({
+      show: () => {
+        setOpen(true);
+        setSelectedEntityType(undefined);
+        setEntityList([]);
+      },
+    }),
+    [],
+  );
+
+  const handleEntityTypeClick = async (entityType: EntityType) => {
+    if (!graphService) return;
+
+    const { entityTypeId } = entityType;
+
+    // get the entities of clicked entity type
+    const { data } = await graphService.aggregateEntities({
+      data: { operation: { entityTypeId, itemsPerPage: 100 } },
+    });
+
+    if (!data) return handleClose();
+
+    const { results: entities } = data;
+
+    // by default, all entity checkboxes are checked
+    const initialSelection = entities.map(() => true);
+
+    setEntityList(entities);
+    setSelections(initialSelection);
+    setSelectedEntityType(entityType);
+  };
+
+  const handleCheckboxChange = (changedIndex: number) => {
+    setSelections((prev) =>
+      prev.map((value, index) => (index === changedIndex ? !value : value)),
+    );
+  };
+
+  const handleAddClick = async () => {
+    if (!graphService) return;
+
+    // create links for selected entities
+    const selectedEntityIds: string[] = [];
+    selections.forEach((value, index) => {
+      if (!value) return;
+
+      const entity = entityList[index];
+      if (entity) selectedEntityIds.push(entity.entityId);
+    });
+
+    const createLinkPromises = selectedEntityIds.map((entityId) =>
+      graphService.createLink({
+        data: {
+          sourceEntityId: blockEntityId,
+          destinationEntityId: entityId,
+          path: "$.items",
+        },
+      }),
+    );
+
+    const createLinkResponses = await Promise.all(createLinkPromises);
+
+    // add the new items to the shuffle items list
+    onAddEntityItems(
+      selectedEntityIds.map((entityId, i) => {
+        return {
+          id: uuid(),
+          value: "",
+          entityId,
+          linkId: createLinkResponses[i]?.data?.linkId,
+        };
+      }),
+    );
+
+    // close the modal
+    handleClose();
+  };
+
+  const selectedItemCount = selections.filter(Boolean).length;
```

Comment:
Hmm, afaik this is widely used in lots of places, and type casting like `String`, `Boolean`, `Number` always accept one argument, which is the first argument. So it should be pretty safe to do so right? We even have this in various places in our codebases. Unless there is a concern about readability or safety that I'm not aware of, I would prefer to keep it this way. Let me know what do you think please 😊

Reply:
Passing a function by name to map which is not specifically designed to be map is an anti pattern.It may work now but you're going to shoot yourself in the foot in the long run. Even if there are exceptions it is easier to avoid this problem by just always avoiding this pattern.

https://jakearchibald.com/2021/function-callback-risks/

Reply:
Hmm, I didn't think there would be such a possibility to JS updating `Boolean()` function to take a second parameter, since it's just a type casting function. Anyhow, updated here https://github.com/hashintel/hash/pull/1033/commits/d5fa504e25a93c1451b3fe018b1f05fdef79d35f